cpp/dolfinx/common/IndexMap.cpp

```diff
@@ -54,6 +54,118 @@ void local_to_global_impl(
+    for (std::int32_t local_index : forward_indices)
+    {
+      for (std::int32_t i = 0; i < bs; ++i)
+      {
+        // Insert field index, global index, composite global index
+        indices.push_back(f);
+        indices.push_back(bs * local_index + i + offset);
+        indices.push_back(bs * local_index + i + local_offset[f]
+                          + process_offset);
+      }
+    }
+  }
+
+  // Build array of neighbourhood ranks
+  std::set<std::int32_t> neighbour_set;
+  for (const common::IndexMap& map : maps)
+  {
+    const std::vector<std::int32_t>& n = map.neighbours();
+    neighbour_set.insert(n.begin(), n.end());
+  }
+  const std::vector<int> neighbours(neighbour_set.begin(), neighbour_set.end());
+
+  // Create neighbourhood communicator
+  MPI_Comm comm;
+  MPI_Dist_graph_create_adjacent(maps.at(0).get().mpi_comm(), neighbours.size(),
+                                 neighbours.data(), MPI_UNWEIGHTED,
+                                 neighbours.size(), neighbours.data(),
+                                 MPI_UNWEIGHTED, MPI_INFO_NULL, false, &comm);
+
+  int num_neighbours(-1), outdegree(-2), weighted(-1);
+  MPI_Dist_graph_neighbors_count(comm, &num_neighbours, &outdegree, &weighted);
+  assert(num_neighbours == outdegree);
+
+  // Figure out how much data to receive from each neighbour
+  const int num_my_rows = indices.size();
+  std::vector<int> num_rows_recv(num_neighbours);
+  MPI_Neighbor_allgather(&num_my_rows, 1, MPI_INT, num_rows_recv.data(), 1,
+                         MPI_INT, comm);
+
+  // Compute displacements for data to receive
+  std::vector<int> disp(num_neighbours + 1, 0);
+  std::partial_sum(num_rows_recv.begin(), num_rows_recv.end(),
+                   disp.begin() + 1);
+
+  // Send data to neighbours, and receive data
+  std::vector<std::int64_t> data_recv(disp.back());
+  MPI_Neighbor_allgatherv(indices.data(), indices.size(), MPI_INT64_T,
+                          data_recv.data(), num_rows_recv.data(), disp.data(),
+                          MPI_INT64_T, comm);
+
+  // Destroy communicator
+  MPI_Comm_free(&comm);
+
+  // Create map (old global index -> new global index) for each field
+  std::vector<std::map<int64_t, std::int64_t>> ghost_maps(maps.size());
+  for (std::size_t i = 0; i < data_recv.size(); i += 3)
+    ghost_maps[data_recv[i]].insert({data_recv[i + 1], data_recv[i + 2]});
+
+  /// Build arrays from old ghost index to composite ghost index for
+  /// each field
+  std::vector<std::vector<std::int64_t>> ghosts_new(maps.size());
+  for (std::size_t f = 0; f < maps.size(); ++f)
+  {
+    const int bs = maps[f].get().block_size();
+    const Eigen::Array<std::int64_t, Eigen::Dynamic, 1>& ghosts
+        = maps[f].get().ghosts();
+    for (Eigen::Index i = 0; i < ghosts.rows(); ++i)
+    {
+      for (int j = 0; j < bs; ++j)
+      {
+        auto it = ghost_maps[f].find(bs * ghosts[i] + j);
+        if (it == ghost_maps[f].end())
+          throw std::runtime_error("Oooops");
```

Comment:
Do we need a more informative error?
Or maybe an assertion, since it's an internal construction.